packages/EDA/wasm/softmax.h

```diff
@@ -0,0 +1,154 @@
+		 learningRate - learning rate
+		 penalty - the penalty coefficient for regularization
+		 tolerance - the fitting tolerance
+		 params - parameters of the model (output of computations)
+
+	   REMARK. Each target value is an integer presenting index category from 0 to classesCount-1.
+	 */
+	template<typename Float, typename Int>
+	int fitSoftmax(Float * features, Float * avgs, Float * stdevs, Int * target, 
+		const int samplesCount, const int featuresCount, const int classesCount,
+		int iterCount, Float learningRate, Float penalty, float tolerance,
+		Float * params) noexcept {
+		
+		// Operating counts
+		const int n = featuresCount;
+		const int m = samplesCount;
+		const int c = classesCount;
+
+        // Train data matrix
+		Matrix<Float, Dynamic, Dynamic, ColMajor> X(n, m);
+        
+		// Transposed train data matrix, used when computaing gradients
+		Map<Matrix<Float, Dynamic, Dynamic, ColMajor>> XT(features, m, n);
+
+        // Vectors with features stats
+		Map<Vector<Float, Dynamic>> avgX(avgs, n);
+		Map<Vector<Float, Dynamic>> stdDevX(stdevs, n);
+
+        // Matrix of train labels in the one-hot form
+		Matrix<Float, Dynamic, Dynamic, ColMajor> Y(c, m);
+		
+		// Parameters of the model to be trained
+		Matrix<Float, Dynamic, Dynamic, RowMajor> W(c, n);
+		Vector<Float, Dynamic> B(c);
+
+        // Operating items
+		Matrix<Float, Dynamic, Dynamic, ColMajor> Z(c, m);		
+		Matrix<Float, Dynamic, Dynamic, RowMajor> dZ(c, m);
+		Matrix<Float, Dynamic, Dynamic, RowMajor> dW(c, n);
+		Vector<Float, Dynamic> dB(c);
+		Vector<Float, Dynamic> sums(m);
+		uint32_t* classWeights = new uint32_t[c];
+		Float loss = 0;
+		Float lossPrev = 0;
+
+		// 1. Prepare data
+
+		// 1.1) Center features
+		XT = XT.rowwise() - avgX.transpose();
+
+		// 1.2) Scale features
+		for (int i = 0; i < n; ++i)
+			if (stdevs[i] > static_cast<Float>(0))
+				XT.col(i) = XT.col(i) / stdDevX(i);
+
+		// 1.3) Transpose
+		X = XT.transpose();
+		
+		// 1.4) One-hot labels & class weights
+		Y = Matrix<Float, Dynamic, Dynamic, ColMajor>::Zero(c, m); 
+		for (int i = 0; i < c; ++i)
+			classWeights[i] = 0;
+        
+		for (int i = 0; i < m; ++i) {
+			Y(target[i], i) = static_cast<Float>(1);
+			classWeights[target[i]] += 1;
+		}
+
+		// 2. Init parameters	
+		Float xavierScale = sqrt(6.0 / (c + n));
+		W = Matrix<Float, Dynamic, Dynamic, RowMajor>::Random(c, n) * xavierScale;
+		B = Vector<Float, Dynamic>::Zero(c);
+		
+		// 3. The main training loop
+		for (int iter = 0; iter < iterCount; ++iter) {			
+			// Z = W * X + B
+			Z = ((W * X).colwise() + B);
+
+            // Z = exp(Z) with weights
+			for (int j = 0; j < m; ++j)
+				for (int i = 0; i < c; ++i)
+					Z(i, j) = exp(Z(i, j)) * classWeights[i];
+			
+			sums = Z.colwise().sum();
+
+            // Compute predicted probabilities
+			for (int i = 0; i < m; ++i)
+				Z.col(i) = Z.col(i) / sums(i);			
+
+            // Copmute loss
+            loss = 0;
```

Comment:
is everything okay with indentation here? was concerned about coomments as well but here is is a line of code that is out of its scope